internal/blocktx/processor.go

```diff
@@ -525,27 +476,97 @@ func (ph *Processor) processBlock(msg *p2p.BlockMessage) {
+	if err = p.markBlockAsProcessed(ctx, block); err != nil {
+		_, errDel := p.store.DelBlockProcessing(ctx, &blockHash, p.hostname)
 		if errDel != nil {
-			ph.logger.Error("failed to delete block processing - after marking block as processed failed", slog.String("hash", blockHash.String()), slog.String("err", errDel.Error()))
+			p.logger.Error("failed to delete block processing - after marking block as processed failed", slog.String("hash", blockHash.String()), slog.String("err", errDel.Error()))
 		}
-		ph.logger.Error("unable to mark block as processed", slog.String("hash", blockHash.String()), slog.String("err", err.Error()))
+		p.logger.Error("unable to mark block as processed", slog.String("hash", blockHash.String()), slog.String("err", err.Error()))
 		return
 	}
 
 	// add the total block processing time to the stats
-	ph.logger.Info("Processed block", slog.String("hash", blockHash.String()), slog.Int("txs", len(msg.TransactionHashes)), slog.String("duration", time.Since(timeStart).String()))
+	p.logger.Info("Processed block", slog.String("hash", blockHash.String()), slog.Int("txs", len(msg.TransactionHashes)), slog.String("duration", time.Since(timeStart).String()))
+}
+
+func (p *Processor) getPrevBlock(ctx context.Context, prevHash, blockHash *chainhash.Hash) (*blocktx_api.Block, error) {
+	prevBlock, err := p.store.GetBlock(ctx, prevHash)
+
+	if err != nil && !errors.Is(err, store.ErrBlockNotFound) {
+		_, errDel := p.store.DelBlockProcessing(ctx, blockHash, p.hostname)
+		if errDel != nil {
+			p.logger.Error("failed to delete block processing - after getting previous block failed", slog.String("hash", blockHash.String()), slog.String("err", errDel.Error()))
+		}
+		return nil, err
+	}
+
+	return prevBlock, nil
+}
+
+func (p *Processor) competingChainsExist(ctx context.Context, block *blocktx_api.Block) (bool, error) {
+	if block.Status == blocktx_api.Status_ORPHANED {
+		return false, nil
+	}
+
+	if block.Status == blocktx_api.Status_LONGEST {
+		competingBlock, err := p.store.GetBlockByHeight(ctx, block.Height, blocktx_api.Status_LONGEST)
+		if err != nil && !errors.Is(err, store.ErrBlockNotFound) {
+			blockHash := (*chainhash.Hash)(block.Hash)
+			_, errDel := p.store.DelBlockProcessing(ctx, blockHash, p.hostname)
+			if errDel != nil {
+				p.logger.Error("failed to delete block processing - after checking for competing chains", slog.String("hash", blockHash.String()), slog.String("err", errDel.Error()))
+			}
+			return false, err
+		}
+
+		if competingBlock != nil && !bytes.Equal(competingBlock.Hash, block.Hash) {
+			return true, nil
+		}
+
+		return false, nil
+	}
+
+	// If STALE status
+	return true, nil
+}
+
+func (p *Processor) hasGreatestChainwork(ctx context.Context, incomingBlock *blocktx_api.Block) (bool, error) {
+	tip, err := p.store.GetChainTip(ctx)
+	if err != nil && !errors.Is(err, store.ErrBlockNotFound) {
+		blockHash := (*chainhash.Hash)(incomingBlock.Hash)
```

Comment:
Thinking if it wouldn't be better to use the chainhash function here as the conversion could fail and it is not checked here🤔 

```
blockHash, err := chainhash.NewHash(block.Hash)
```

Reply:
Changed